projects/packages/connection/src/class-manager.php

```diff
@@ -785,43 +739,8 @@ public function connect_user( $user_id = null, $redirect_url = null ) {
 	 * @param bool    $can_overwrite_primary_user Allow for the primary user to be disconnected.
 	 * @return Boolean Whether the disconnection of the user was successful.
 	 */
-	public static function disconnect_user( $user_id = null, $can_overwrite_primary_user = false ) {
-		$tokens = Jetpack_Options::get_option( 'user_tokens' );
-		if ( ! $tokens ) {
-			return false;
-		}
-
-		$user_id = empty( $user_id ) ? get_current_user_id() : (int) $user_id;
-
-		if ( Jetpack_Options::get_option( 'master_user' ) === $user_id && ! $can_overwrite_primary_user ) {
-			return false;
-		}
-
-		if ( ! isset( $tokens[ $user_id ] ) ) {
-			return false;
-		}
-
-		$xml = new \Jetpack_IXR_Client( compact( 'user_id' ) );
-		$xml->query( 'jetpack.unlink_user', $user_id );
-
-		unset( $tokens[ $user_id ] );
-
-		Jetpack_Options::update_option( 'user_tokens', $tokens );
-
-		// Delete cached connected user data.
-		$transient_key = "jetpack_connected_user_data_$user_id";
-		delete_transient( $transient_key );
-
-		/**
-		 * Fires after the current user has been unlinked from WordPress.com.
-		 *
-		 * @since 4.1.0
-		 *
-		 * @param int $user_id The current user's ID.
-		 */
-		do_action( 'jetpack_unlinked_user', $user_id );
-
-		return true;
+	public function disconnect_user( $user_id = null, $can_overwrite_primary_user = false ) {
```

Comment:
I don't think that gets us towards scalability because it requires reading all tokens and writing all tokens in each of those proposed API endpoints. What is needed is methods that only read or write single tokens for single users.

What I did do, though, is moved more of the Tokens method into the Manager method, so that it's responsible for clearing caches and defaulting to the current user. This seems like a nice simplification of the Tokens class.

Reply:
> I don't think that gets us towards scalability because it requires reading all tokens and writing all tokens in each of those proposed API endpoints. What is needed is methods that only read or write single tokens for single users.

I agree. My point was mostly that if the logic for updating tokens is inside Tokens, then it will be easier to optimize it in the future any way we'd choose.

> What I did do, though, is moved more of the Tokens method into the Manager method, so that it's responsible for clearing caches and defaulting to the current user. This seems like a nice simplification of the Tokens class.

Awesome, thanks!